### PR TITLE
Respect Docker HEALTHCHECK(s) when starting sidecar containers

### PR DESCRIPTION
**Pull request by dakale**
I wrote a sample that uses Postgres as a sidecar container. When the container starts, it takes ~10 seconds before the database will actually respond to requests, which is shorter than the fastest time it takes to start a job that pings the database, leading to a race condition error

Docker has HEALTHCHECK as a first class feature. This simply uses `docker inspect` to check if sidecar containers have a healthcheck, and if so it delays the "Initialize containers" step until the container is truly ready (healthcheck status == `healthy`), by whatever definition of ready it provides as part of the container

### CODE REVIEW

File: src/Agent.Worker/ContainerOperationProvider.cs [hunk 1]
```diff
@@ -524,5 +529,32 @@ private async Task RemoveContainerNetworkAsync(IExecutionContext executionContex
...
+            if (string.IsNullOrEmpty(serviceHealth))
+            {
+                // Container has no HEALTHCHECK
+                return;
+            }
+            TimeSpan backoff = TimeSpan.FromSeconds(2);
+            while (string.Equals(serviceHealth, "starting", StringComparison.OrdinalIgnoreCase))
+            {
+                executionContext.Output($"{container.ContainerNetworkAlias} service is starting, waiting {backoff.Seconds} seconds before checking again.");
+                Thread.Sleep(backoff);
```
Comment (TingluoHuang) [inline src/Agent.Worker/ContainerOperationProvider.cs]: never use Thread.Sleep() in async/await Task. :D
try use `BackoffTimerHelper` to get timespan.

Comment (TingluoHuang) [reply src/Agent.Worker/ContainerOperationProvider.cs]: you can find `BackoffTimerHelper ` in this file. :)

File: src/Agent.Worker/ContainerOperationProvider.cs [hunk 2]
```diff
@@ -128,6 +128,11 @@ public async Task StartContainersAsync(IExecutionContext executionContext, objec
             {
                 await StartContainerAsync(executionContext, container);
             }
+
+            foreach (var container in containers.FindAll(c => !c.IsJobContainer))
```
Comment (TingluoHuang) [inline src/Agent.Worker/ContainerOperationProvider.cs]: why not just `.Where()` instead of `FindAll`?

**Comment by dakale** [comment]
@TingluoHuang I think all the comments should be addressed now

**Pull request merged**